### ISSUE TITLE
Hazelcast node continuosly logs WrongTargetException

### ISSUE THREAD

**Issue by agherardi** [hazelcast/hazelcast#3395]
Hi,
While running a load test, after a few hours one of the nodes logs thousand of warnings like this:

2014-08-26T06:20:51.898Z] [WARN] [-] [hz._hzInstance_1_vmscore_1.operation.thread-2] [com.hazelcast.spi.impl.WaitNotifyServiceImpl$WaitingOp] - [192.168.240.25]:5701 [vmscore_1] [3.2] Op: com.hazelcast.concurrent.lock.operations.LockOperation@4a642c96, com.hazelcast.spi.exception.WrongTargetException: WrongTarget! this:Address[192.168.240.25]:5701, target:Address[192.168.240.24]:5701, partitionId: 194, replicaIndex: 0, operation: com.hazelcast.spi.impl.WaitNotifyServiceImpl$WaitingOp, service: hz:impl:lockService

We're running 3.2 so I believe we have picked up the patch for 
https://github.com/hazelcast/hazelcast/issues/1815

Any ideas?

**Comment by SimonTurfer** [comment]
We're getting this exception on version 3.3.3 as well, running on EC2 instances. It happens when one instance is being shutdown from the AWS console.

We're also getting spammed with "Retrying[2] packet send operation to...." messages when this happens and it wont stop retrying.

```
[hz._hzInstance_1_dev.cached.thread-22] 2015-01-04 18:03:09,928 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-22] 2015-01-04 18:03:09,932 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[1] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-22] 2015-01-04 18:03:09,937 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-22] 2015-01-04 18:03:09,953 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-22] 2015-01-04 18:03:09,954 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-22] 2015-01-04 18:03:09,955 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-22] 2015-01-04 18:03:09,956 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-3] 2015-01-04 18:03:09,956 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-3] 2015-01-04 18:03:09,958 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-3] 2015-01-04 18:03:09,964 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-3] 2015-01-04 18:03:09,969 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-6] 2015-01-04 18:03:09,969 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
[hz._hzInstance_1_dev.cached.thread-3] 2015-01-04 18:03:09,970 [x.x.x.x]:5701 [dev] [3.3.3] Retrying[2] packet send operation to: Address[x.x.x.x]:5701
```

**Comment by innokenty-shuvalov** [comment]
Same for 3.5, thousands of such entries were logged, although the cluster looks to be working okay in general:
```
Jul 06, 2015 6:00:01 PM com.hazelcast.spi.impl.waitnotifyservice.impl.WaitingOperation
WARNING: [hz-aerostat-test02.haze.yandex.net]:5701 [aerostat-testing] [3.5] Op: com.hazelcast.concurrent.lock.operations.LockOperation{serviceName='null', partitionId=199, callId=0, invocationTime=1436195180735, waitTimeout=240000, callTimeout=60000}, com.hazelcast.spi.exception.WrongTargetException: WrongTarget! this:Address[hz-aerostat-test02.haze.yandex.net]:5701, target:Address[hz-aerostat-test01.haze.yandex.net]:5701, partitionId: 199, replicaIndex: 0, operation: com.hazelcast.spi.impl.waitnotifyservice.impl.WaitingOperation, service: hz:impl:lockService
```

**Comment by wenerme** [comment]
After WrongTargetException, the my data get lost,how about your's ?

3.5.2

```
23:52:11 [main] WARN  com.hazelcast.logging.Slf4jFactory$Slf4jLogger:90 - [127.0.0.1]:5702 [dev] [3.5.2] Exception while getting future
java.util.concurrent.ExecutionException: com.hazelcast.spi.exception.WrongTargetException: WrongTarget! this:Address[127.0.0.1]:5703, target:Address[127.0.0.1]:5701, partitionId: 139, replicaIndex: 1, operation: com.hazelcast.partition.impl.IsReplicaVersionSync, service: hz:core:partitionService
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveApplicationResponseOrThrowException(InvocationFuture.java:357) ~
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:225) ~
	at com.hazelcast.partition.impl.InternalPartitionServiceImpl.getFutureResult(InternalPartitionServiceImpl.java:1124) 
	at com.hazelcast.partition.impl.InternalPartitionServiceImpl.isReplicaInSyncState(InternalPartitionServiceImpl.java:1108) 
	at com.hazelcast.partition.impl.InternalPartitionServiceImpl.getMemberState(InternalPartitionServiceImpl.java:1049) 
	at com.hazelcast.partition.impl.InternalPartitionServiceImpl.isMemberStateSafe(InternalPartitionServiceImpl.java:1035) 
	at com.hazelcast.partition.impl.SafeStateCheckOperation.run(SafeStateCheckOperation.java:35) 
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:137) 
	at com.hazelcast.spi.impl.operationexecutor.classic.ClassicOperationExecutor.runOnCallingThread(ClassicOperationExecutor.java:380) 
	at com.hazelcast.spi.impl.operationexecutor.classic.ClassicOperationExecutor.runOnCallingThreadIfPossible(ClassicOperationExecutor.java:342) 
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:247) 
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:231) 
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invokeInternal(Invocation.java:198) 
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:168) 
	at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.invokeOnTarget(OperationServiceImpl.java:277) 
	at com.hazelcast.partition.PartitionServiceProxy.isClusterSafe(PartitionServiceProxy.java:107) 
	at com.ServerServerStartup.start(GiServerStartup.java:68) 
	at com.ServerServerStartup.main(GiServerStartup.java:64) 
Caused by: com.hazelcast.spi.exception.WrongTargetException: WrongTarget! this:Address[127.0.0.1]:5703, target:Address[127.0.0.1]:5701, partitionId: 139, replicaIndex: 1, operation: com.hazelcast.partition.impl.IsReplicaVersionSync, service: hz:core:partitionService
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.ensureNoPartitionProblems(OperationRunnerImpl.java:246) ~
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:127) ~
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:315) ~
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:142) ~
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:115) ~
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:101) ~
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:76) ~
	at ------ End remote and begin local stack-trace ------.(Unknown Source) ~[?:?]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveApplicationResponse(InvocationFuture.java:384) ~
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveApplicationResponseOrThrowException(InvocationFuture.java:334) ~
	... 17 more
```

**Comment by peterlitvak** [comment]
Getting this on 3.5.2, in EC2 when doing rolling restart of the nodes.
One thing is that our app calls map.lock() in one of our beans upon the instantiation.

**Comment by agherardi** [comment]
I submitted a tentative patch with this pull request: https://github.com/hazelcast/hazelcast/pull/7825

**Comment by colinduplantis** [comment]
I have seen this problem as well with 3.5.x. For me, it requires some kind of cluster-wide operation (getLock or executeOnAllMembers, eg) while an instance is starting up. FWIW, this problem seems easier to reproduce on a 16-core machine.

**Comment by tehsu** [comment]
I'm seeing this as well, do we know if there has been a fix, it looks like the PR is waiting still.

**Comment by colinduplantis** [comment]
I don't have any info on a fix yet.

It's very difficult for us to reproduce reliably. One thing that has 
popped up for us, though, is that we were using Oracle as our RDBMS. 
There is a bug/feature in the Oracle JDBC driver that causes it to 
consume a fair bit of randomness from /dev/random on start. Combine this 
with using a pre-3.16 Linux kernel which was slow to collect entropy and 
a bug in Oracle Java 7 configuration that incorrectly sets the random 
generator to the blocking random device instead of the unblocking random 
device.

Now, none of that has anything to do with Hazelcast yet, except, we 
found that Hazelcast _seems_ to be sensitive to executing cluster-wide 
operations (lock, exec-service execution on all members, replicating 
hash map, etc) during start.

So, in a scenario where half our cluster has started and is executing, 
and another set of machines joins the party late, the first set is more 
likely to execute these cluster-wide operations while the second set is 
starting up. There appears to be a brief window during cluster start 
where Hcast is vulnerable, with a deadlock resulting if a cluster-wide 
operation is executed at exactly the right time during start.

Finally, the Oracle thing comes in. We start our cluster, which starts 
the Hcast service via Spring config, which depends on our DB pool. The 
DB pool fires up and Hcast via Spring will wait to complete 
initialization until the DB pool is started. The DB pool is blocking on 
/dev/random because of the clusterfsck detailed above. This causes Hcast 
to be held in a state in which it appears to be vulnerable to deadlock 
for seconds to minutes, thus inviting disaster.

Factor in that we've only managed to reproduce this on a 16-core server.

This may or may not be what you're seeing, and we still can't quite nail 
it down, but we haven't seen a recurrence since we applied a tweak to 
force the JVM to use the unblocking random device.

The fix is to add this to your JVM at start:

-Djava.security.egd=file:/dev/../dev/urandom

Of course, none of this may apply to you.

On 06/24/2016 08:28 AM, Tehseen wrote:
>
> I'm seeing this as well, do we know if there has been a fix, it looks 
> like the PR is waiting still.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/hazelcast/hazelcast/issues/3395#issuecomment-228376849>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AJIdsu5W0D2_UNlLN5a9o8sl9dtYAlhUks5qO_eUgaJpZM4CbUnB>.
>

-- 
Colin DuPlantis
Chief Architect, Marketcetera
Download, Run, Trade
888.868.4884 +1.541.306.6556
http://www.marketcetera.org

**Comment by bradmac** [comment]
We've just seen the same issue in EC2 after a rolling deploy

**Comment by peterlitvak** [comment]
Yes, we had to abandon hazelcast for this reason in our current app unfortunately.

**Comment by jerrinot** [comment]
Related to #3754

**Comment by tombujok** [comment]
Thanks for your valuable comments guys. We have a fix on the way as @jerrinot pointed out. 
We're planning to address it in the forthcoming days.

**Comment by jerrinot** [comment]
fixed by #8550

**Comment by raftAtGit** [comment]
not sure if the same issue, but I also see a continuous stream of similar logs sometimes:
```
INFO: [192.168.178.108]:5702 [dev] [4.0] Cannot execute leadership rebalance at the moment: java.util.concurrent.CompletionException: com.hazelcast.spi.exception.TargetNotMemberException: Not Member! target: [192.168.178.108]:5701, partitionId: -1, operation: com.hazelcast.cp.internal.operation.GetLeadedGroupsOp, service: hz:core:raft
May 08, 2020 10:37:19 PM com.hazelcast.cp.internal.RaftGroupMembershipManager
INFO: [192.168.178.108]:5702 [dev] [4.0] Cannot execute leadership rebalance at the moment: java.util.concurrent.CompletionException: com.hazelcast.spi.exception.TargetNotMemberException: Not Member! target: [192.168.178.108]:5701, partitionId: -1, operation: com.hazelcast.cp.internal.operation.GetLeadedGroupsOp, service: hz:core:raft
May 08, 2020 10:38:19 PM com.hazelcast.cp.internal.RaftGroupMembershipManager
INFO: [192.168.178.108]:5702 [dev] [4.0] Cannot execute leadership rebalance at the moment: java.util.concurrent.CompletionException: com.hazelcast.spi.exception.TargetNotMemberException: Not Member! target: [192.168.178.108]:5701, partitionId: -1, operation: com.hazelcast.cp.internal.operation.GetLeadedGroupsOp, service: hz:core:raft
May 08, 2020 10:39:19 PM com.hazelcast.cp.internal.RaftGroupMembershipManager
INFO: [192.168.178.108]:5702 [dev] [4.0] Cannot execute leadership rebalance at the moment: java.util.concurrent.CompletionException: com.hazelcast.spi.exception.TargetNotMemberException: Not Member! target: [192.168.178.108]:5701, partitionId: -1, operation: com.hazelcast.cp.internal.operation.GetLeadedGroupsOp, service: hz:core:raft
```
in this last case, there were 4 nodes, with 3 CP members and 3 CP group size, I killed one node and one of the nodes started logging this (it was the Raft leader I guess)

and the system just hanged, possibly at `IAtomicLong.incrementAndGet`

version: `com.hazelcast:hazelcast:4.0`

**Comment by mmedenjak** [comment]
@raftAtGit no, it's not the same issue. I suggest opening a new issue. Maybe @mdogan or @petrpleshachkov know something more based on your message above.
